### PR TITLE
Fix state hash mismatch error that caused by ignored MPT error

### PR DESCRIPTION
**Pull request by peterlimg**
This PR was created to fix the new detected 'state hash mismatch' error that caused the chain stuck. But after that, new issues were detected during the loadtests, hence all fixes were also added here. Anyway, the core fixes and changes are as shown below.

## Fixes
- Fixed the state hash mismatch error that caused the chain stuck. 'node not found' error was not returned to upper layer that caused the txn executor see the error as charageable error, while other miners that has full state would not see this error, therefore, they will see different block state, and cause chain stuck.
- Fixed the slow finalizing process on sharders, it took about 1.2s to store the miner rewards update events to db. Optimized to do bulk update for all the events in one sql execution. Also, maximize the concurrency on finalizing a blocks. 
- Fixed the partition save bug
- Fixed a sharder blocks sync bug that caused the sharder stuck
- Fixed allocation challenges stats bug

## Changes
- Changed the stake pools to use `txn.ClientID` as key. Previously, we use txn hash as key, which does not make sense at all, as each provider has limited size of delegate/stake pools, if we use txn hash as key, them we can only have max limit size of txns to stake the pools. For example, if miner one has max of 200 delegate pools, if we use txn hash as pool key, then we will only allow 200 stake transactions, then no more staking could be allowed. While, after the fix, we use client id as pool id, which means, as long as the client can stake the miner once, it then can stake as many times as it wants, and new staked tokens will all be added up to the same pool. This would also simplify the clients usage, basically all the pool_id params will be removed from the `zboxcli`, `zwalletcli`.
- Minor refactoring for fixing potential goroutine leaks.

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- gosdk: https://github.com/0chain/gosdk/pull/556
- system_test: https://github.com/0chain/system_test/pull/380
- zboxcli: https://github.com/0chain/zboxcli/pull/291
- zwalletcli: https://github.com/0chain/zwalletcli/pull/172
- loadTest: https://github.com/0chain/loadTest/pull/53

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/chaincore/chain/protocol_block.go [hunk 1]
```diff
@@ -403,26 +418,62 @@ func (c *Chain) finalizeBlock(ctx context.Context, fb *block.Block, bsh BlockSta
...
+		c.DeleteBlocks(deadBlocks)
+	})
+
+	wg.Wait()
+	logging.Logger.Debug("finalized block - done",
+		zap.Int64("round", fb.Round), zap.String("block", fb.Hash),
+		zap.Any("duration", time.Since(ts)))
+}
+
+type waitGroupSync struct {
```
Comment (dabasov) [inline code/go/0chain.net/chaincore/chain/protocol_block.go:447]: we might want to mow it to common.context? it is reasonable to have concurrency structures all in one place

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 2]
```diff
@@ -1279,21 +1315,32 @@ func (sc *StorageSmartContract) finalizedPassRates(alloc *StorageAllocation) ([]
...
 			ba.Stats = new(StorageAllocationStats)
+			passRates = append(passRates, 1.0)
+			continue
 		}
 		ba.Stats.FailedChallenges += ba.Stats.OpenChallenges
 		ba.Stats.OpenChallenges = 0
-		ba.Stats.TotalChallenges = ba.Stats.FailedChallenges + ba.Stats.SuccessChallenges
-		if ba.Stats.TotalChallenges == 0 {
+
+		baTotal := ba.Stats.FailedChallenges + ba.Stats.SuccessChallenges
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:1347]: should we update ``ba.Stats.TotalChallenges`` with this value?

**Review by dabasov — APPROVED** [review]
lgtm

File: code/go/0chain.net/smartcontract/storagesc/allocation.go (hunk 2 above)
Comment (peterlimg) [reply code/go/0chain.net/smartcontract/storagesc/allocation.go:1347]: No, the `ba.Stats.TotalChallenges` will only be updated when new challenges are generated, update here could lead to wrong stats.

File: code/go/0chain.net/chaincore/chain/protocol_block.go (hunk 1 above)
Comment (peterlimg) [reply code/go/0chain.net/chaincore/chain/protocol_block.go:447]: Perhaps, if needed, we can move to a common package.

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | staging  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | one-pool  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**